### PR TITLE
[scan][trainer] hotfix for xcresulttool spec changes >= XC16b3

### LINKED ISSUE #22132: Xcode 16 beta 3 tests fail as xcresulttool get object is deprecated
### New Issue Checklist

- [ yes] Updated fastlane to the latest version
- [ yes] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [ yes] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [ yes] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

This doesn't work anymore
```bash
/Applications/Xcode-beta-16-v3.app/Contents/Developer/usr/bin/xcresulttool get --format json --path /fastlane/test_output/Test-MyAppTests-2024.07.09_14-57-50-339029-+0700.xcresult        
Error: This command is deprecated and will be removed in a future release, --legacy flag is required to use it.
Usage: xcresulttool get object [--legacy] --path <path> [--id <id>] [--version <version>] [--format <format>]
  See 'xcresulttool get object --help' for more information.
```

Seems quick fix maybe done by adding _--legacy_ flag.

Thank you.

##### Command executed
```ruby
run_tests(...)
```

##### Complete output when running fastlane, including the stack trace and command used

```bash
[15:28:35]: ▸ Test execute Succeeded
Error: This command is deprecated and will be removed in a future release, --legacy flag is required to use it.
Usage: xcresulttool get object [--legacy] --path <path> [--id <id>] [--version <version>] [--format <format>]
  See 'xcresulttool get object --help' for more information.

Failed to execute - xcrun xcresulttool get --format json --path fastlane/test_output/Test-MyAppTests-2024.07.09_15-27-46-331299-+0700.xcresult

vendor/bundle/ruby/3.1.0/gems/fastlane-2.221.1/trainer/lib/trainer/test_parser.rb:196:in `execute_cmd': \e[31m[!] Failed to execute - xcrun xcresulttool get --format json --path fastlane/test_output/Test-MyAppTests-2024.07.09_15-27-46-331299-+0700.xcresult\e[0m (RuntimeError)
    from vendor/bundle/ruby/3.1.0/gems/fastlane-2.221.1/trainer/lib/trainer/test_parser.rb:205:in `parse_xcresult'
    from vendor/bundle/ruby/3.1.0/gems/fastlane-2.221.1/trainer/lib/trainer/test_parser.rb:105:in `initialize'
    from vendor/bundle/ruby/3.1.0/gems/fastlane-2.221.1/trainer/lib/trainer/test_parser.rb:82:in `new'
```

### Environment

Xcode 16 beta 3
MacOS 14.5 (23F79)
Fastlane 2.221.1
ruby 3.1.0

### PR DESCRIPTION
**Pull request by dokimyj**
### Checklist

- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
  - for *trainer* module, all passed
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I see several green `ci/circleci` builds in the "All checks have passed" section of my PR ([connect CircleCI to GitHub](https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-) if not)
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [ ] I've updated the documentation if necessary.
  - N/A
- [x] I've added or updated relevant unit tests.

### Motivation and Context

Resolves https://github.com/fastlane/fastlane/issues/22132

From Xcode 16 beta 3, `fastlane ios scan` started to fail after xctest had finished. 
The reason is: There has been a deprecation and stderr in traditional `xcrun xcresulttool` command usage.

<details>
  <summary> Error details of xcresulttool related portion </summary>

```
...
00:07:00 Error: This command is deprecated and will be removed in a future release, --legacy flag is required to use it.
00:07:00 Usage: xcresulttool get object [--legacy] --path <path> [--id <id>] [--version <version>] [--format <format>]
00:07:00   See 'xcresulttool get object --help' for more information.
...
```

</details>

I tried new way when obtaining xcresult object (`xcresulttool get tests` or something) but
- test summary was separated
- the entire format of the object was totally different
- test ID for `--id` option was not showing up in the report

For these reasons, decided to add `--legacy` if using Xcode 16 beta 3. 
It would take quite a time to use the new type of `xcresulttool` like in the deprecation notice.

### Description

used changes of this branch as a *fastlane* gem and tried `run_tests` under XC15 and 16.
specs in *trainer* had nothing to test with this part so I tried *scan* on both Xcode 15.4 and 16 beta 3

### Testing Steps

```
00:00:16 $ gem uninstall -Iax fastlane
00:00:16 git clone --single-branch -b dokimyj-patch-1 https://github.com/dokimyj/fastlane fastlane-dokimyj
00:00:16 bundle install --path=vendor/bundle
```

Both environments are in CI, I cloned this single branch and `bundle install` to use bundle fastlane.

```
00:06:53 ▸ Testing passed on 'iPhone 15 Plus'
00:06:53 ▸ Test Succeeded
00:06:57 Skipping HTML... only available with `xcodebuild_formatter: 'xcpretty'` right now
00:06:57 Your 'xcodebuild_formatter' doesn't support these 'output_types'. Change your 'output_types' to prevent these warnings from showing...
00:06:57 +-------------------------+
00:06:57 |      Test Results       |
00:06:57 +--------------------+----+
00:06:57 | Number of tests    | 12 |
00:06:57 | Number of failures | 0  |
00:06:57 +--------------------+----+
```

*scan* with Xcode 16 beta 3 using this branch

```
00:08:36 ▸ Testing passed on 'iPhone 15 Plus'
00:08:36 ▸ Test Succeeded
00:08:41 Skipping HTML... only available with `xcodebuild_formatter: 'xcpretty'` right now
00:08:41 Your 'xcodebuild_formatter' doesn't support these 'output_types'. Change your 'output_types' to prevent these warnings from showing...
00:08:42 +-------------------------+
00:08:42 |      Test Results       |
00:08:42 +--------------------+----+
00:08:42 | Number of tests    | 12 |
00:08:42 | Number of failures | 0  |
00:08:42 +--------------------+----+
```

*scan* with Xcode 15.4 using this branch

Seems no differences in the result!

### CODE REVIEW

File: trainer/lib/trainer/test_parser.rb [hunk 1]
```diff
@@ -202,7 +202,11 @@ def parse_xcresult(path, output_remove_retry_attempts: false)
       path = Shellwords.escape(path)
 
       # Executes xcresulttool to get JSON format of the result bundle object
-      result_bundle_object_raw = execute_cmd("xcrun xcresulttool get --format json --path #{path}")
+      # Hotfix: From Xcode 16 beta 3 'xcresulttool get --format json' has been deprecated; '--legacy' flag required to keep on using the command
+      xcresulttool_cmd = "xcrun xcresulttool get --format json --path #{path}"
+      xcresulttool_cmd << ' --legacy' if `xcrun --version`.gsub('xcrun version ', '').to_f >= 70.0 # Version of xcrun bundled in Xcode 16 beta 3
```
Comment (giginet) [inline trainer/lib/trainer/test_parser.rb:207]: Could you use `xcrun xcresulttool version` instead to check backward compatibility?
This behavior depends on `xcresulttool` version, not `xcrun` version.

`xcresulttool version 23021, format version 3.53 (current)`

`format version` can be used.

It's better to extract the method to parse `xcresulttool` version

```ruby
private
def xcresulttool_version
  result.match(/format version (?<format_version>\d+\.\d+)/)
end
```

File: trainer/lib/trainer/test_parser.rb [hunk 2]
```diff
@@ -202,7 +202,11 @@ def parse_xcresult(path, output_remove_retry_attempts: false)
       path = Shellwords.escape(path)
 
       # Executes xcresulttool to get JSON format of the result bundle object
-      result_bundle_object_raw = execute_cmd("xcrun xcresulttool get --format json --path #{path}")
+      # Hotfix: From Xcode 16 beta 3 'xcresulttool get --format json' has been deprecated; '--legacy' flag required to keep on using the command
+      xcresulttool_cmd = "xcrun xcresulttool get --format json --path #{path}"
```
Comment (giginet) [inline trainer/lib/trainer/test_parser.rb:206]: I prefer to use an array to construct args

```ruby
args = %W(
  xcrun
  xcresulttool
  get
  --format
  json
  --path
  #{path}
)
args << "--legacy" if xcresulttool_version >= 3.53
execute_cmd(args.join(' '))
```

**Review by giginet — COMMENTED** [review]
Thank you!

It's better to add specs on https://github.com/fastlane/fastlane/blob/master/trainer/spec/test_parser_spec.rb with stubbing

**Comment by giginet** [comment]
I saw the spec implementation; writing tests for these changes is difficult, so we may skip writing specs.

File: trainer/lib/trainer/test_parser.rb (hunk 1 above)
Comment (dokimyj) [reply trainer/lib/trainer/test_parser.rb:207]:
```
$ DEVELOPER_DIR=/Applications/Xcode_15.4.app xcrun xcresulttool version
xcresulttool version 22608, format version 3.49 (current)

$ DEVELOPER_DIR=/Applications/Xcode_16_beta.app xcrun xcresulttool version
xcresulttool version 23019.2, format version 3.53 (current)

$ DEVELOPER_DIR=/Applications/Xcode_16_beta_2.app xcrun xcresulttool version
xcresulttool version 23019.2, format version 3.53 (current)

$ DEVELOPER_DIR=/Applications/Xcode_16_beta_3.app xcrun xcresulttool version
xcresulttool version 23021, format version 3.53 (current)
```

Thanks for the information. I have never known there is `xcresulttool version` command and just accepted suggestion from other. 
Pinpointing `xcresulttool` version would be the best answer. Thanks again!

File: trainer/lib/trainer/test_parser.rb [hunk 3]
```diff
@@ -197,12 +197,36 @@ def execute_cmd(cmd)
...
+        xcresulttool
+        get
+        --format
+        json
+        --path
+        #{path}
+      )
+
+      # Version of xcresulttool bundled in Xcode 16 beta 3 was 23021
+      xcresulttool_version_string = `xcrun xcresulttool version`.split(',')[0]
```
Comment (giginet) [inline trainer/lib/trainer/test_parser.rb:214]: Could you note the expected output from `xcrun xcresulttool version` on the comment?

```ruby
# xcresulttool version 23021, format version 3.53 (current)
```

Comment (giginet) [reply trainer/lib/trainer/test_parser.rb:214]: I prefer to use regex and the named capture for code readability

```ruby
match = result.match(/xcresulttool version (?<version>\d+)/)
version = match[:version]&.to_f
```

File: trainer/spec/test_parser_spec.rb [hunk 4]
```diff
@@ -26,6 +26,24 @@
...
+      end
+
+      it "use traditional command for parse_xcresult on < Xcode 16 beta 3", requires_xcode: true do
+        xcresult_sample_path = "./trainer/spec/fixtures/Test.test_result.xcresult"
+        @model = Trainer::TestParser.new(xcresult_sample_path)
+        allow_any_instance_of(Trainer::TestParser).to receive(:`).and_return("xcresulttool version 22608, format version 3.49 (current)")
+        actual = @model.send(:generate_cmd_parse_xcresult, xcresult_sample_path)
+        expect(actual).to eq("xcrun xcresulttool get --format json --path #{xcresult_sample_path}")
+      end
+    end
```
Comment (giginet) [inline trainer/spec/test_parser_spec.rb:45]: It's better to unify test logic with `context`

```suggestion
    describe "#generate_cmd_parse_xcresult" do
      let(:xcresult_sample_path) { "./trainer/spec/fixtures/Test.test_result.xcresult" }
      let(:subject) { Trainer::TestParser.new(xcresult_sample_path) }
      
      before do
        allow_any_instance_of(Trainer::TestParser).to receive(:`).and_return(version)
      end
      
      context 'with >= Xcode 16 beta 3' do
        let(:version) { 'xcresulttool version 23021, format version 3.53 (current)' }
      
        it 'should pass `--legacy`' do
          actual = subject.send(:generate_cmd_parse_xcresult, xcresult_sample_path)
          expect(actual).to eq("xcrun xcresulttool get --format json --path #{xcresult_sample_path} --legacy")
        end
      end

      context 'with < Xcode 16 beta 3' do
        let(:version) { 'xcresulttool version 22608, format version 3.49 (current)' }
      
        it 'should not pass `--legacy`' do
          actual = subject.send(:generate_cmd_parse_xcresult, xcresult_sample_path)
          expect(actual).to eq("xcrun xcresulttool get --format json --path #{xcresult_sample_path}")
        end
      end
    end
```

Using `context` to describe the situation is the best practice of rspec.

**Review by giginet — COMMENTED** [review]
Thank you for fixing the PR.

It almost seems to be good.

File: trainer/spec/test_parser_spec.rb (hunk 4 above)
Comment (giginet) [reply trainer/spec/test_parser_spec.rb:45]: Sorry, my suggestion was wrong.

I confirmed this block works well.

```ruby
    describe "#generate_cmd_parse_xcresult" do
      let(:xcresult_sample_path) { "./trainer/spec/fixtures/Test.test_result.xcresult" }
      let(:json_content) { JSON.load(absolute_path) }
      let!(:subject) { Trainer::TestParser.new(xcresult_sample_path) }

      before do
        allow(File).to receive(:expand_path).with(xcresult_sample_path).and_return(xcresult_sample_path)
        allow_any_instance_of(Trainer::TestParser).to receive(:`).with('xcrun xcresulttool version').and_return(version)
      end

      context 'with >= Xcode 16 beta 3' do
        let(:version) { 'xcresulttool version 23021, format version 3.53 (current)' }
        let(:expected) { "xcrun xcresulttool get --format json --path #{xcresult_sample_path} --legacy" }

        it 'should pass `--legacy`' do
          actual = subject.send(:generate_cmd_parse_xcresult, xcresult_sample_path)
          expect(actual).to eq(expected)
        end
      end

      context 'with < Xcode 16 beta 3' do
        let(:version) { 'xcresulttool version 22608, format version 3.49 (current)' }
        let(:expected) { "xcrun xcresulttool get --format json --path #{xcresult_sample_path}" }

        it 'should not pass `--legacy`' do
          actual = subject.send(:generate_cmd_parse_xcresult, xcresult_sample_path)
          expect(actual).to eq(expected)
        end
      end
    end
```

We have to use `let!` to define the subject before stubbing the backquote.

My point is using `context` to follow the [better Rspec](https://www.betterspecs.org/#contexts)

**Comment by dokimyj** [comment]
# Amended along with the review comments:

- Separated method for `xcrun xcresulttool get` command generation (for Rspec test :) )
- Judging from `xcresulttool version`, not `xcrun --version`
  - Reviewer recommended `format version` condition, but from Xcode 16 beta it has never changed in beta 3.
- Constructing argument with `%W` instead of oneliner string
- Added Rspec, formatted
- Some writing style fixes

# How I checked if the fixes work:

I tried on my local machine using Xcode 15.4, Xcode 16 beta, and Xcode 16 beta 3.

- Xcode 15.4
```
$ DEVELOPER_DIR=/Applications/Xcode_15.4.app bundle exec fastlane ios test
[✔] 🚀

...

| xcode_path                                     | /Applications/Xcode_15.4.app                                      |
+------------------------------------------------+-------------------------------------------------------------------+

...

[09:55:36]: ▸ Test Succeeded
+------------------------+
|      Test Results      |
+--------------------+---+
| Number of tests    | 6 |
| Number of failures | 0 |
+--------------------+---+

...
```

- Xcode 16 beta

```
$ DEVELOPER_DIR=/Applications/Xcode_16_beta.app bundle exec fastlane ios test
[✔] 🚀

...

| xcode_path                                     | /Applications/Xcode_16_beta.app                                   |
+------------------------------------------------+-------------------------------------------------------------------+

...

[09:59:38]: ▸ Test Succeeded
+------------------------+
|      Test Results      |
+--------------------+---+
| Number of tests    | 6 |
| Number of failures | 0 |
+--------------------+---+
...

```

- Xcode 16 beta 3

```
$ DEVELOPER_DIR=/Applications/Xcode_16_beta_3.app bundle exec fastlane ios test
[✔] 🚀

...

| xcode_path                                     | /Applications/Xcode_16_beta_3.app                                 |
+------------------------------------------------+-------------------------------------------------------------------+

...

[10:18:47]: ▸ Test Succeeded
+------------------------+
|      Test Results      |
+--------------------+---+
| Number of tests    | 6 |
| Number of failures | 0 |
+--------------------+---+

...
```

And... would anyone please restart *appveyor/pr* to see if CI is green? I got no permission to start that job I think...

File: trainer/spec/test_parser_spec.rb [hunk 5]
```diff
@@ -26,6 +26,35 @@
...
+      before do
+        allow(File).to receive(:expand_path).with(xcresult_sample_path).and_return(xcresult_sample_path)
+        allow_any_instance_of(Trainer::TestParser).to receive(:`).with('xcrun xcresulttool version').and_return(version)
+      end
+
+      context 'with >= Xcode 16 beta 3' do
+        let(:version) { 'xcresulttool version 23021, format version 3.53 (current)' }
+        let(:expected) { "xcrun xcresulttool get --format json --path #{xcresult_sample_path} --legacy" }
+
+        it 'should pass `--legacy`', requires_xcode: true do
```
Comment (giginet) [inline trainer/spec/test_parser_spec.rb:43]: What is `requires_xcode`?

**Review by giginet — APPROVED** [review]
Thank your continuous effort. Looks great for me 🎉 

I'll resolve CI error and I'll merge this soon.

Thank you again!

File: trainer/spec/test_parser_spec.rb (hunk 5 above)
Comment (giginet) [reply trainer/spec/test_parser_spec.rb:43]: OK, it's better to add this tag.

Comment (giginet) [reply trainer/spec/test_parser_spec.rb:43]: In this case, we only check the command builder. It doesn't depend on Xcode, so we can get rid of this.

Comment (dokimyj) [reply trainer/spec/test_parser_spec.rb:43]: if not adding `requires_xcode` flag, ubuntu test will fail on CI.

https://app.circleci.com/pipelines/github/fastlane/fastlane/6614/workflows/3db274a0-fb41-4ce4-9783-a50fda4ff9c2/jobs/101945?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary

```
Failures:

  1) Trainer Trainer::TestParser#generate_cmd_parse_xcresult with >= Xcode 16 beta 3 should pass `--legacy`
     Failure/Error: let!(:subject) { Trainer::TestParser.new(xcresult_sample_path) }

     Errno::ENOENT:
       No such file or directory - xcrun
```

**Comment by giginet** [comment]
Great work! Thank you!

**Pull request merged**